File: core/src/main/java/com/github/aecsocket/minecommons/core/effect/Effector.java
package com.github.aecsocket.minecommons.core.effect;

import com.github.aecsocket.minecommons.core.vector.cartesian.Vector3;

/**
 * An audience-like object which can accept an effect.
 */
public interface Effector {
    /**
     * Gets an effector that does nothing.
     * @return The effector.
     */
    static Effector empty() {
        return EmptyEffector.INSTANCE;
    }

    /**
     * Plays a sound effect, with a precomputed distance.
     * @param effect The sound effect.
     * @param origin The origin position of the sound.
     * @param distance If already computed, the distance between the effector and the origin.
     */
    void play(SoundEffect effect, Vector3 origin, double distance);

    /**
     * Plays a sound effect.
     * @param effect The sound effect.
     * @param origin The origin position of the sound.
     */
    void play(SoundEffect effect, Vector3 origin);

    /**
     * Spawns a particle effect.
     * @param effect The particle effect.
     * @param origin The origin position of the particle.
     */
    void spawn(ParticleEffect effect, Vector3 origin);
}

File: core/src/main/kotlin/com/github/aecsocket/alexandria/core/effect/Effector.kt
package com.github.aecsocket.alexandria.core.effect

import com.github.aecsocket.alexandria.core.spatial.Vector3

interface Effector {
    fun playSound(effect: SoundEffect, position: Vector3)

    fun showParticle(effect: ParticleEffect, position: Vector3)

    companion object {
        val EMPTY: Effector = EmptyEffector
    }
}

private object EmptyEffector : Effector {
    override fun playSound(effect: SoundEffect, position: Vector3) {}
    override fun showParticle(effect: ParticleEffect, position: Vector3) {}
}

fun interface ForwardingEffector : Effector {
    fun effectors(): Iterable<Effector>

    override fun playSound(effect: SoundEffect, position: Vector3) =
        effectors().forEach { it.playSound(effect, position) }

    override fun showParticle(effect: ParticleEffect, position: Vector3) =
        effectors().forEach { it.showParticle(effect, position) }
}


